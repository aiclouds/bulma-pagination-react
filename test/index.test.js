import React from 'react';

import renderer from 'react-test-renderer';

// eslint-disable-next-line unicorn/import-index
import Pagination, { Page, getVisiblePages } from '../index';

const render = element => renderer.create(element).toJSON();

describe('Page', () => {
  test('should display a numbered element', () => {
    expect(
      render(<Page currentPage={1} index={19} onChange={() => {}} />)
    ).toMatchSnapshot();
  });

  test('should be different for matching current', () => {
    expect(
      render(<Page currentPage={1} index={1} onChange={() => {}} />)
    ).toMatchSnapshot();
  });

  test('should use className', () => {
    expect(
      renderer
        .create(
          <Page
            currentPage={1}
            index={1}
            className="foo-bar"
            onChange={() => {}}
          />
        )
        .root.findByProps({ className: 'foo-bar' })
    ).toBeDefined();
  });

  test('should use onChange callback', () => {
    const onChange = jest.fn();
    const tree = renderer.create(
      <Page currentPage={1} index={1} onChange={onChange} />
    );
    tree.root.find(e => e.type === 'a').props.onClick();
    expect(onChange).toHaveBeenCalledWith(1);
  });
});

describe('getVisiblePages', () => {
  test('at start of list', () => {
    expect(getVisiblePages(1, 1, 10)).toEqual([1, 2, 3]);
  });

  test('at end of list', () => {
    expect(getVisiblePages(1, 10, 10)).toEqual([8, 9, 10]);
  });

  test('in the middle of a list', () => {
    expect(getVisiblePages(1, 5, 10)).toEqual([4, 5, 6]);
  });

  test('less pages than radius', () => {
    expect(getVisiblePages(1, 1, 2)).toEqual([1, 2]);
  });
});

describe('Pagination', () => {
  test('should render pages at start', () => {
    expect(
      render(<Pagination pages={10} onChange={() => {}} />)
    ).toMatchSnapshot();
  });

  test('should render pages at start with many', () => {
    expect(
      render(<Pagination pages={10} currentPage={4} visibleRadius={2} />)
    ).toMatchSnapshot();
  });

  test('should render pages in middle', () => {
    expect(
      render(<Pagination pages={10} currentPage={5} onChange={() => {}} />)
    ).toMatchSnapshot();
  });

  test('should render pages at end', () => {
    expect(
      render(<Pagination pages={10} currentPage={10} onChange={() => {}} />)
    ).toMatchSnapshot();
  });

  test('should render with no ellipses if not needed - visible == max', () => {
    expect(
      render(<Pagination pages={3} currentPage={1} onChange={() => {}} />)
    ).toMatchSnapshot();
  });

  test('should render with no ellipses if not needed - end', () => {
    expect(
      render(<Pagination pages={10} currentPage={8} onChange={() => {}} />)
    ).toMatchSnapshot();
  });

  test('should render with no ellipses if not needed - start', () => {
    expect(
      render(<Pagination pages={10} currentPage={3} onChange={() => {}} />)
    ).toMatchSnapshot();
  });

  test('should call callback when previous clicked', () => {
    const onChange = jest.fn();
    const tree = renderer.create(
      <Pagination currentPage={1} pages={10} onChange={onChange} />
    );
    tree.root
      .find(
        e => e.type === 'a' && e.props.className.includes('pagination-previous')
      )
      .props.onClick();
    expect(onChange).toHaveBeenCalledWith(0);
  });

  test('should call callback when next clicked', () => {
    const onChange = jest.fn();
    const tree = renderer.create(
      <Pagination currentPage={1} pages={10} onChange={onChange} />
    );
    tree.root
      .find(
        e => e.type === 'a' && e.props.className.includes('pagination-next')
      )
      .props.onClick();
    expect(onChange).toHaveBeenCalledWith(2);
  });
});
