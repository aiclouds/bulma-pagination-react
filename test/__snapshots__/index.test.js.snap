// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Page should be different for matching current 1`] = `
<li>
  <a
    aria-current="page"
    aria-label="Goto page 1"
    className="pagination-link  is-current"
    onClick={[Function]}
  >
    1
  </a>
</li>
`;

exports[`Page should display a numbered element 1`] = `
<li>
  <a
    aria-current={false}
    aria-label="Goto page 19"
    className="pagination-link  false"
    onClick={[Function]}
  >
    19
  </a>
</li>
`;

exports[`Pagination should render pages at end 1`] = `
<nav
  aria-label="pagination"
  className="pagination-container pagination  "
  role="navigation"
>
  <a
    className="pagination-previous nav "
    disabled={false}
    onClick={[Function]}
  >
    Previous
  </a>
  <a
    className="pagination-next nav "
    disabled={true}
    onClick={[Function]}
  >
    Next page
  </a>
  <ul
    className="pagination-list "
  >
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 1"
        className="pagination-link  false"
        onClick={[Function]}
      >
        1
      </a>
    </li>
    <li>
      <span
        className="pagination-ellipsis"
      >
        …
      </span>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 8"
        className="pagination-link  false"
        onClick={[Function]}
      >
        8
      </a>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 9"
        className="pagination-link  false"
        onClick={[Function]}
      >
        9
      </a>
    </li>
    <li>
      <a
        aria-current="page"
        aria-label="Goto page 10"
        className="pagination-link  is-current"
        onClick={[Function]}
      >
        10
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Pagination should render pages at start 1`] = `
<nav
  aria-label="pagination"
  className="pagination-container pagination  "
  role="navigation"
>
  <a
    className="pagination-previous nav "
    disabled={true}
    onClick={[Function]}
  >
    Previous
  </a>
  <a
    className="pagination-next nav "
    disabled={false}
    onClick={[Function]}
  >
    Next page
  </a>
  <ul
    className="pagination-list "
  >
    <li>
      <a
        aria-current="page"
        aria-label="Goto page 1"
        className="pagination-link  is-current"
        onClick={[Function]}
      >
        1
      </a>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 2"
        className="pagination-link  false"
        onClick={[Function]}
      >
        2
      </a>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 3"
        className="pagination-link  false"
        onClick={[Function]}
      >
        3
      </a>
    </li>
    <li>
      <span
        className="pagination-ellipsis"
      >
        …
      </span>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 10"
        className="pagination-link  false"
        onClick={[Function]}
      >
        10
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Pagination should render pages at start with many 1`] = `
<nav
  aria-label="pagination"
  className="pagination-container pagination  "
  role="navigation"
>
  <a
    className="pagination-previous nav "
    disabled={false}
    onClick={[Function]}
  >
    Previous
  </a>
  <a
    className="pagination-next nav "
    disabled={false}
    onClick={[Function]}
  >
    Next page
  </a>
  <ul
    className="pagination-list "
  >
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 1"
        className="pagination-link  false"
        onClick={[Function]}
      >
        1
      </a>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 2"
        className="pagination-link  false"
        onClick={[Function]}
      >
        2
      </a>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 3"
        className="pagination-link  false"
        onClick={[Function]}
      >
        3
      </a>
    </li>
    <li>
      <a
        aria-current="page"
        aria-label="Goto page 4"
        className="pagination-link  is-current"
        onClick={[Function]}
      >
        4
      </a>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 5"
        className="pagination-link  false"
        onClick={[Function]}
      >
        5
      </a>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 6"
        className="pagination-link  false"
        onClick={[Function]}
      >
        6
      </a>
    </li>
    <li>
      <span
        className="pagination-ellipsis"
      >
        …
      </span>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 10"
        className="pagination-link  false"
        onClick={[Function]}
      >
        10
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Pagination should render pages in middle 1`] = `
<nav
  aria-label="pagination"
  className="pagination-container pagination  "
  role="navigation"
>
  <a
    className="pagination-previous nav "
    disabled={false}
    onClick={[Function]}
  >
    Previous
  </a>
  <a
    className="pagination-next nav "
    disabled={false}
    onClick={[Function]}
  >
    Next page
  </a>
  <ul
    className="pagination-list "
  >
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 1"
        className="pagination-link  false"
        onClick={[Function]}
      >
        1
      </a>
    </li>
    <li>
      <span
        className="pagination-ellipsis"
      >
        …
      </span>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 4"
        className="pagination-link  false"
        onClick={[Function]}
      >
        4
      </a>
    </li>
    <li>
      <a
        aria-current="page"
        aria-label="Goto page 5"
        className="pagination-link  is-current"
        onClick={[Function]}
      >
        5
      </a>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 6"
        className="pagination-link  false"
        onClick={[Function]}
      >
        6
      </a>
    </li>
    <li>
      <span
        className="pagination-ellipsis"
      >
        …
      </span>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 10"
        className="pagination-link  false"
        onClick={[Function]}
      >
        10
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Pagination should render with no ellipses if not needed - end 1`] = `
<nav
  aria-label="pagination"
  className="pagination-container pagination  "
  role="navigation"
>
  <a
    className="pagination-previous nav "
    disabled={false}
    onClick={[Function]}
  >
    Previous
  </a>
  <a
    className="pagination-next nav "
    disabled={false}
    onClick={[Function]}
  >
    Next page
  </a>
  <ul
    className="pagination-list "
  >
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 1"
        className="pagination-link  false"
        onClick={[Function]}
      >
        1
      </a>
    </li>
    <li>
      <span
        className="pagination-ellipsis"
      >
        …
      </span>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 7"
        className="pagination-link  false"
        onClick={[Function]}
      >
        7
      </a>
    </li>
    <li>
      <a
        aria-current="page"
        aria-label="Goto page 8"
        className="pagination-link  is-current"
        onClick={[Function]}
      >
        8
      </a>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 9"
        className="pagination-link  false"
        onClick={[Function]}
      >
        9
      </a>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 10"
        className="pagination-link  false"
        onClick={[Function]}
      >
        10
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Pagination should render with no ellipses if not needed - start 1`] = `
<nav
  aria-label="pagination"
  className="pagination-container pagination  "
  role="navigation"
>
  <a
    className="pagination-previous nav "
    disabled={false}
    onClick={[Function]}
  >
    Previous
  </a>
  <a
    className="pagination-next nav "
    disabled={false}
    onClick={[Function]}
  >
    Next page
  </a>
  <ul
    className="pagination-list "
  >
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 1"
        className="pagination-link  false"
        onClick={[Function]}
      >
        1
      </a>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 2"
        className="pagination-link  false"
        onClick={[Function]}
      >
        2
      </a>
    </li>
    <li>
      <a
        aria-current="page"
        aria-label="Goto page 3"
        className="pagination-link  is-current"
        onClick={[Function]}
      >
        3
      </a>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 4"
        className="pagination-link  false"
        onClick={[Function]}
      >
        4
      </a>
    </li>
    <li>
      <span
        className="pagination-ellipsis"
      >
        …
      </span>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 10"
        className="pagination-link  false"
        onClick={[Function]}
      >
        10
      </a>
    </li>
  </ul>
</nav>
`;

exports[`Pagination should render with no ellipses if not needed - visible == max 1`] = `
<nav
  aria-label="pagination"
  className="pagination-container pagination  "
  role="navigation"
>
  <a
    className="pagination-previous nav "
    disabled={true}
    onClick={[Function]}
  >
    Previous
  </a>
  <a
    className="pagination-next nav "
    disabled={false}
    onClick={[Function]}
  >
    Next page
  </a>
  <ul
    className="pagination-list "
  >
    <li>
      <a
        aria-current="page"
        aria-label="Goto page 1"
        className="pagination-link  is-current"
        onClick={[Function]}
      >
        1
      </a>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 2"
        className="pagination-link  false"
        onClick={[Function]}
      >
        2
      </a>
    </li>
    <li>
      <a
        aria-current={false}
        aria-label="Goto page 3"
        className="pagination-link  false"
        onClick={[Function]}
      >
        3
      </a>
    </li>
  </ul>
</nav>
`;
